### Bin&auml;r-, Oktal, Dezimal- und Hexadezimalzahlen
Das Dezimalsystem verwendet die Ziffern 0 bis 9, und es gilt

$$1234 = 1 \cdot 1000 + 2 \cdot 100 + 3 \cdot 10 + 4\cdot 1 =
  1 \cdot 10^3 + 2 \cdot 10^2 + 3 \cdot 10^1 + 4\cdot 10^0.$$  

Das Bin&auml;rsystem verwendet entsprechend nur die Ziffern 0 und 1.
Der Pr&auml;fix 0b zeigt an, dass eine Zahl als bin&auml;re Zahl zu interpretieren ist,
also

$$0b1011 = 1 \cdot 2^3 +  0 \cdot 2^2 + 1 \cdot 2^1 + 1\cdot 2^0 = 11$$

als Dezimalzahl.  

Das Oktalsystem verwendet die Ziffern 0 is 7, und der Pr&auml;fix 0o zeigt an, 
dass eine Zahl als oktale Zahl zu interpretieren ist, also
 
$$ 0o1234 = 1\cdot 8^3+2\cdot 8^2+3\cdot 8 + 4 = 512+2*64+3\cdot 8 +4 = 668.$$


Das Hexadezimalsystem verwendet die Zeichen 0,1,...,9,A,B,C,D,E,F, wobei A,..,F f&uuml;r 10,...,15 stehen. Oft werden auch a,...f anstelle von A,...,F verwendet.
Der Pr&auml;fix 0x zeigt an, 
dass eine Zahl als  Hexadezimalzahl zu interpretieren ist, also z.B.

$$ 0xCAFE = 12\cdot 16^3+10\cdot 16^2+15\cdot 16 + 14 = 51966.$$

In [167]:
x = 0b1011
y = 0o1234
z = 0xCAFE
print(x,y,z, sep = ', ')

11, 668, 51966


***
Zahl als Bin&auml;r-, Oktal- oder Hexadezimalzahl ausgegen:
***

In [170]:
print('{:b}'.format(11))
print('{:o}'.format(668))
print('{:x}'.format(51966))

1011
1234
cafe


***
Wieso verwechseln Informatiker Helloween mit Weihnachten?
***

In [173]:
0o31 == 25 # 31oct(tal) gleich 25 dez(imal)

True

### Bitoperationen  
(siehe auch [hier](https://python-reference.readthedocs.io/en/latest/docs/operators/#bitwise-operators))  
Die Operatoren `&` (bitweises **and**) `|` (bitweises **or**) und `^` (bitweises 
**xor**) verkn&uuml;pfen zwei Zahlen bitweise: Das i-te Bit des Resultats ist
das Resultat der entsprechenden Verkn&uuml;pfung der i-ten Bits der beiden Zahlen.

```
1100 &              1100 |            1100 ^  
1010                1010              1010 
----                ----              ----
1000                1110              0110 
```

Die bitshift Operatoren `<<` und `>>` schieben die Bits in der Bin&auml;rdarstellung 
nach link oder rechts, z.B. `1011 << 2 = 101100` und `1011 >> 2 = 10`.

Um zu testen ob das i-te Bit einer Zahl `x` gesetzt ist,
verwendet man die Operation `1 << i & x`: Das i-te Bit ist gesetzt, falls dieser
Ausdruch ungleich 0 ist, genauer, gleich `1 << i`.
Z.B. ist bei der Zahl `11=0b1011` das 2. Bit gesetzt, das 1. Bit jedoch nicht (wobei das rechteste Bit das 0. Bit ist).
 

```
1011 &          1011 &
1000            0100
----            ----
1000 = 1<<3        0
```

In [183]:
# '{:04b}': gib Zahl als Binaerzahl aus. 
# Fuege fuehrende Nullen hinzu falls weniger als 4 ZIffern

print('{:04b}'.format(0b1100 & 0b1010))
print('{:04b}'.format(0b1100 | 0b1010))
print('{:04b}'.format(0b1100 ^ 0b1010))

1000
1110
0110


In [187]:
print(1 << 3 & 11)
print(1 << 3 & 11 == 1 << 3)
print(1 << 2 & 11)

8
True
0


In [18]:
def is_ith_bit_set(i, n):
    'return 1<<i if ith bit is set else 0'
    return 1 << i & n

In [189]:
is_ith_bit_set(2, 15)

True

In [197]:
def dec2bin(n):
    'ith digit is 1 if ith bit is set, else 0'
    digits = {True: '1', False: '0'}
    res = ''
    for i in range(n.bit_length()):
        ith_bit = bool(is_ith_bit_set(i, n))
        digit = digits[ith_bit]
        res += digit
    return res     

In [196]:
dec2bin(11)

'1101'

In [3]:
'{:04b}'.format(11)

'1011'